lib/galaxy/tools/__init__.py

```diff
@@ -2378,6 +2378,20 @@ def check_dataset_instance(input_dataset):
 
             map(check_dataset_instance, input_dataset_collection.dataset_instances)
 
+    def _add_datasets_to_history(self, history, elements_dict_or_list):
+        datasets = []
+        if isinstance(elements_dict_or_list, list):
+            for element_object in elements_dict_or_list:
+                if getattr(element_object, "history_content_type", None) == "dataset":
+                    datasets.append(element_object)
+        else:
+            for element_object in elements_dict_or_list.values():
```

Comment:
Wouldn't it be cleaner if this method would accept only lists, i.e. we pass `.values()` of the dict instead of the dict?
Or start this method with `if not isinstance(elements_dict_or_list, list): elements_dict_or_list = elements_dict_or_list.values()`

Reply:
Yes, good point - I'll fix this up.